[release-4.15] OCPBUGS-28928: Prevent upgrades for SHA1 default cert and SHA1 route certs

gcs278 opened the pull request:
_**Note: This is NOT a backport or cherry-pick of any commits, but instead a discrete PR targeting 4.15.**_

This PR handles SHA1 certificate upgradeable scenarios for 4.15. With OpenSSL3.0 provided by RHEL9 in 4.16, the router will fail to start if any cert provided is SHA1.

First, if the default certificate on the Ingress Controller object is using SHA1, then set `Upgradeable` to be False.

Secondly, add a new control loop, route deprecated, that determines upgradeablity by searching for the `UnservableInFutureVersions` condition in the routes. The `UnservableInFutureVersions` will be added to the route by https://github.com/openshift/router/pull/555 when a SHA1 certificate exists on the route. It creates an [admin-gate](https://github.com/openshift/enhancements/blob/master/enhancements/update/upgrades-blocking-on-ack.md) if a `UnservableInFutureVersions` condition is found. This logic assumes any `UnservableInFutureVersions` route status condition is an upgrade blocker.

This implementation targets 4.15 to 4.16 updates specifically, so it is only targeting the `release-4.15` branch and we will be merged in the backport of https://issues.redhat.com/browse/OCPBUGS-26498. The code should not be merged into 4.16.

This PR leveraged logic from https://github.com/openshift/vmware-vsphere-csi-driver-operator/pull/171.

This PR depends on https://github.com/openshift/router/pull/585 for E2E test to succeed.

Miciah commented on pkg/operator/controller/route-deprecated/controller.go:
```diff
@@ -0,0 +1,227 @@
+	// Index routes over the deprecated condition so that we can
+	// efficiently find deprecated routes.
+	if err := reconciler.cache.IndexField(context.Background(), &routev1.Route{}, "deprecatedRoute", client.IndexerFunc(func(o client.Object) []string {
+		if cond := getRouteDeprecated(o.(*routev1.Route)); cond != nil {
```
Please name the function slightly more descriptively:
```suggestion
		if cond := getRouteDeprecatedStatusCondition(o.(*routev1.Route)); cond != nil {
```

gcs278 replied:
Done.

Miciah commented on pkg/operator/controller/route-deprecated/controller.go:
```diff
@@ -0,0 +1,227 @@
+			return configmap.Name == adminGateConfigMapName && configmap.Namespace == operatorcontroller.GlobalMachineSpecifiedConfigNamespace
+		},
+	}
+
+	if err != nil {
+		return nil, err
+	}
```
Please move the blank line.  
```suggestion
	if err != nil {
		return nil, err
	}

```

gcs278 replied:
Oh this was completely in the wrong area. Moved.

Miciah commented on pkg/operator/controller/route-deprecated/controller.go:
```diff
@@ -0,0 +1,227 @@
+		return nil, err
+	}
+	// add watch for changes for Routes
+	if err := c.Watch(source.Kind(reconciler.cache, &routev1.Route{}),
+		&handler.EnqueueRequestForObject{},
+		updatedDeprecatedConditionFilter); err != nil {
+		return nil, err
+	}
+	// add watch for changes for ConfigMaps
+	if err := c.Watch(source.Kind(operatorCache, &corev1.ConfigMap{}),
```
Please add a comment explaining why you use one cache for `Watch(source.Kind(reconciler.cache, &routev1.Route{})` but a different cache for `Watch(source.Kind(operatorCache, &corev1.ConfigMap{})`.

gcs278 replied:
Done.

wking commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -716,14 +715,8 @@ func checkDefaultCertificate(secret *corev1.Secret, domain string) error {
-		if cert.Subject.CommonName == domain && !foundSAN {
-			return fmt.Errorf("certificate in secret %s/%s has legacy Common Name (CN) but has no Subject Alternative Name (SAN) for domain: %s", secret.Namespace, secret.Name, domain)
+		if cert.SignatureAlgorithm == x509.SHA1WithRSA || cert.SignatureAlgorithm == x509.ECDSAWithSHA1 {
+			return fmt.Errorf("certificate in secret %s/%s has weak SHA1 signature algorithm: %s", secret.Namespace, secret.Name, domain)
```
Worth a link out to https://docs.openshift.com/container-platform/4.16/release_notes/ocp-4-16-release-notes.html#ocp-4-16-networking ?  And possibly extending that section to fix "HaProxy" -> "HAProxy" and link https://docs.openshift.com/container-platform/4.16/networking/routes/secured-routes.html ?

gcs278 commented:
/retest-required

ShudiLi commented:
/jira refresh

gcs278 commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -716,14 +715,8 @@ func checkDefaultCertificate(secret *corev1.Secret, domain string) error {
-				foundSAN = true
-			}
-		}
-		if cert.Subject.CommonName == domain && !foundSAN {
```
Removing this SAN upgrade logic seems problematic as we are backporting this in 4.15, so then 4.15 will not have this logic, but 4.16+ will have it. 

I think the logic needs to be removed in the master branch instead of removed in a backport like this. The 4.16 bug https://issues.redhat.com/browse/OCPBUGS-26498 is already ON_QA and we've branched to 4.17, so it's probably too late to merge this in master and try to backport it.

I think I should just handle this later when appropriate, keep this backport as "additive" only.

gcs278 replied:
Updated to not remove the SAN logic.

ShudiLi commented:
/retest

ShudiLi commented:
Tested it 4.15.0-0.test-2024-05-29-011612-ci-ln-9c615dk-latest

1.
% oc get clusterversion
NAME      VERSION                                                AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.15.0-0.test-2024-05-29-011612-ci-ln-9c615dk-latest   True        False         86m     Cluster version is 4.15.0-0.test-2024-05-29-011612-ci-ln-9c615dk-latest

2.
 % oc create route edge myedge1 --service=unsec-server3  --cert=testCertificateSHA1.crt --key=testCertificateSHA1.key
route.route.openshift.io/myedge1 created

3.
% oc get route
NAME      HOST/PORT                                                           PATH   SERVICES        PORT   TERMINATION   WILDCARD
myedge1   myedge1-default.apps.shudi-415g29.qe.gcp.devcluster.openshift.com          unsec-server3   http   edge          None

4.
% oc -n openshift-ingress get pods
NAME                            READY   STATUS    RESTARTS   AGE
router-default-cdd85756-9jjmk   1/1     Running   0          25m
router-default-cdd85756-x2vw4   1/1     Running   0          24m

5.
% oc -n openshift-ingress delete pod router-default-cdd85756-9jjmk
pod "router-default-cdd85756-9jjmk" deleted

6.
% oc -n openshift-ingress get pods
NAME                            READY   STATUS    RESTARTS   AGE
router-default-cdd85756-mbtfx   1/1     Running   0          2m5s
router-default-cdd85756-x2vw4   1/1     Running   0          26m

ShudiLi commented:
since the image created by the bot hasn't the signature, the cluster with the image couldn't be updated without --force option, I will do the upgrade test after this PR is merged.

/label qe-approved
thanks

gcs278 commented:
/assign @frobware

ShudiLi commented:
/label qe-approved

ShudiLi commented:
/retest-required

gcs278 commented:
https://github.com/openshift/router/pull/585 has merged, CI should pass now, however I did see [e2e-gcp-operator](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1014/pull-ci-openshift-cluster-ingress-operator-release-4.15-e2e-gcp-operator/1796435275308601344) passed, which is concerning. Testing manually to see if I have a false positive, but going to let operator CI jobs spin:
/test e2e-aws-operator
/test e2e-azure-operator
/test e2e-gcp-operator

gcs278 commented:
`: Find all of the input images from ocp/4.15 and tag them into the stable stream` failure:
/test e2e-gcp-operator

gcs278 commented:
While I figure out false positive E2E test issue.
/hold

gcs278 commented:
some weird infrastructure issues:
/retest

gcs278 commented:
Infrastructure failure
/retest

gcs278 commented:
/test images

gcs278 commented:
/test images

gcs278 commented:
/test images

gcs278 commented:
/test images

gcs278 commented:
/test images

frobware reviewed (changes requested):
I've dropped in a few suggestions. Mostly about identifying config map names in error messages. And various other observations along the way.

gcs278 commented:
/test images

gcs278 commented:
/test images

gcs278 commented on test/e2e/route_upgradeable_test.go:
```diff
@@ -0,0 +1,192 @@
+// waitForClusterVersionUpgradeableCondition waits for an expected upgradeable condition in the CVO.
+// It assumes no upgradeable condition means a cluster is upgradeable.
+func waitForClusterVersionUpgradeableCondition(t *testing.T, expectedStatus v1.ConditionStatus) {
+	err := wait.PollUntilContextTimeout(context.TODO(), 3*time.Second, 3*time.Minute, false, func(ctx context.Context) (bool, error) {
```
Done.

gcs278 commented on test/e2e/route_upgradeable_test.go:
```diff
@@ -0,0 +1,192 @@
+
+	for i := range clusterVersion.Status.Conditions {
+		if clusterVersion.Status.Conditions[i].Type == v1.OperatorUpgradeable {
+			upgradeableCondition = &clusterVersion.Status.Conditions[i]
```
Yes. Done.

gcs278 commented on test/e2e/route_upgradeable_test.go:
```diff
@@ -0,0 +1,192 @@
+	adminAcksConfigMap := &corev1.ConfigMap{}
+	adminAcksNamespacedName := types.NamespacedName{Name: adminAcksConfigMapName, Namespace: operatorcontroller.GlobalUserSpecifiedConfigNamespace}
+	if err := kclient.Get(context.TODO(), adminAcksNamespacedName, adminAcksConfigMap); err != nil {
+		t.Fatalf("failed to get admin-acks configmap: %v", err)
```
Done. I should have cleaned up all of these.

gcs278 reviewed (commented):
Thanks for the review @frobware. 

All comments should be addressed. Deltas: https://github.com/openshift/cluster-ingress-operator/compare/716791955a5d67958f01b2cf619138325e3d4d0a..880737d60a7d0eed769f61a8867a09d3ddb7071d

gcs278 commented:
CI should be sorted:
/unhold

frobware reviewed (changes requested):
A couple of observations which will be trivial to address. Overall LGTM now that the test is vastly simplified. I was able to run the test on my 4.15 cluster.

gcs278 commented on test/e2e/route_upgradeable_test.go:
```diff
@@ -0,0 +1,95 @@
+		}
+	}()
+
+	// Poll	for the admin gate, and wait until it is True.
```
Good catch. Done.

gcs278 commented on test/e2e/util_test.go:
```diff
@@ -843,3 +845,34 @@ func createNamespace(t *testing.T, name string) *corev1.Namespace {
+			if _, ok := adminGateConfigMap.Data[adminGateKey]; ok == expectExists {
+				return true, nil
+			}
+		}
```
Ah yes, Done.

gcs278 reviewed (commented):
Thanks for the review!

frobware commented:
/lgtm

gcs278 commented:
/retest

Miciah commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -725,6 +728,9 @@ func checkDefaultCertificate(secret *corev1.Secret, domain string) error {
 		if cert.Subject.CommonName == domain && !foundSAN {
 			return fmt.Errorf("certificate in secret %s/%s has legacy Common Name (CN) but has no Subject Alternative Name (SAN) for domain: %s", secret.Namespace, secret.Name, domain)
 		}
+		if cert.SignatureAlgorithm == x509.SHA1WithRSA || cert.SignatureAlgorithm == x509.ECDSAWithSHA1 {
```
Using `switch` is slightly more idiomatic:
```suggestion
		switch cert.SignatureAlgorithm {
		case x509.SHA1WithRSA, x509.ECDSAWithSHA1:
```

Miciah commented:
[The ingress-operator logs](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1014/pull-ci-openshift-cluster-ingress-operator-release-4.15-e2e-azure-operator/1799075698002890752/artifacts/e2e-azure-operator/gather-extra/artifacts/pods/openshift-ingress-operator_ingress-operator-644c6994b5-rrmmz_ingress-operator.log) from [the last e2e-azure-operator run](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1014/pull-ci-openshift-cluster-ingress-operator-release-4.15-e2e-azure-operator/1799075698002890752) illustrate the weirdness I mentioned in a couple of my review comments:
```text
% curl -s 'https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1014/pull-ci-openshift-cluster-ingress-operator-release-4.15-e2e-azure-operator/1799075698002890752/artifacts/e2e-azure-operator/gather-extra/artifacts/pods/openshift-ingress-operator_ingress-operator-644c6994b5-rrmmz_ingress-operator.log' | grep -e route_upgradeable_controller
2024-06-07T14:30:58.295Z        INFO    operator.init   controller/controller.go:234    Starting EventSource    {"controller": "route_upgradeable_controller", "source": "kind source: *v1.Route"}
2024-06-07T14:30:58.296Z        INFO    operator.init   controller/controller.go:234    Starting EventSource    {"controller": "route_upgradeable_controller", "source": "kind source: *v1.ConfigMap"}
2024-06-07T14:30:58.296Z        INFO    operator.init   controller/controller.go:234    Starting Controller     {"controller": "route_upgradeable_controller"}
2024-06-07T14:30:58.501Z        INFO    operator.init   controller/controller.go:234    Starting workers        {"controller": "route_upgradeable_controller", "worker count": 1}
2024-06-07T14:30:58.501Z        INFO    operator.route_upgradeable_controller   controller/controller.go:119    reconciling     {"request": {"name":"admin-gates","namespace":"openshift-config-managed"}}
2024-06-07T15:36:47.679Z        INFO    operator.route_upgradeable_controller   controller/controller.go:119    reconciling     {"request": {"name":"route-sha1","namespace":"openshift-ingress"}}
2024-06-07T15:36:47.687Z        INFO    operator.route_upgradeable_controller   route-upgradeable/controller.go:192     Adding admin gate for unservable in future versions route
2024-06-07T15:36:47.698Z        INFO    operator.route_upgradeable_controller   controller/controller.go:119    reconciling     {"request": {"name":"admin-gates","namespace":"openshift-config-managed"}}
2024-06-07T15:36:47.703Z        INFO    operator.route_upgradeable_controller   route-upgradeable/controller.go:192     Adding admin gate for unservable in future versions route
2024-06-07T15:36:50.941Z        INFO    operator.route_upgradeable_controller   controller/controller.go:119    reconciling     {"request": {"name":"route-sha1","namespace":"openshift-ingress"}}
2024-06-07T15:36:50.947Z        INFO    operator.route_upgradeable_controller   route-upgradeable/controller.go:188     Removing admin gate for unservable in future versions route
2024-06-07T15:36:50.958Z        INFO    operator.route_upgradeable_controller   controller/controller.go:119    reconciling     {"request": {"name":"admin-gates","namespace":"openshift-config-managed"}}
% 
```
You can see that the reconciliation requests include both the "admin-gates" configmap and the "route-sha1" route, and you can see that "Adding admin gate for unservable in future versions route" gets logged twice even though (as far as I can tell) the gate only needs to be added once.

gcs278 commented on pkg/operator/controller/route-upgradeable/controller.go:
```diff
@@ -0,0 +1,285 @@
+
+	if _, ok := adminGateConfigMap.Data[RouteConfigAdminGate415Key]; ok {
+		return nil // Already exists.
+	}
```
Good catch. Added a check for the msg value.

gcs278 commented on test/e2e/route_upgradeable_test.go:
```diff
@@ -0,0 +1,95 @@
+// TestRouteNotUpgradeableWithSha1 tests the route-upgradeable control loop by creating a route with a SHA1 certificate,
+// expecting the Ingress Operator to create an admin gate, and update the route to SHA256 certificate to resolve the
+// issue.
+func TestRouteNotUpgradeableWithSha1(t *testing.T) {
```
Correct. Added a comment.

gcs278 commented on test/e2e/route_upgradeable_test.go:
```diff
@@ -0,0 +1,95 @@
+	}
+	certSha1, keySha1, err := generateClientCertificate(ca, caKey, "test", x509.SHA1WithRSA)
+	if err != nil {
+		t.Fatalf("failed to generate SHA1 client certificate: %v", err)
```
Done.

gcs278 commented on test/e2e/route_upgradeable_test.go:
```diff
@@ -0,0 +1,95 @@
+	if err != nil {
+		t.Fatalf("failed to generate client CA certificate: %v", err)
+	}
+	certSha1, keySha1, err := generateClientCertificate(ca, caKey, "test", x509.SHA1WithRSA)
```
Renamed and moved to `util_tls_test.go`.

gcs278 commented on test/e2e/route_upgradeable_test.go:
```diff
@@ -0,0 +1,95 @@
+			},
+			TLS: &routev1.TLSConfig{
+				Termination: routev1.TLSTerminationEdge,
+				Certificate: encodeCert(certSha1),
+				Key:         encodeKey(keySha1),
```
Done.

gcs278 commented on test/e2e/route_upgradeable_test.go:
```diff
@@ -0,0 +1,95 @@
+	if err := kclient.Create(context.Background(), routeSha1); err != nil {
+		t.Fatalf("failed to create route %s/%s: %v", routeSha1.Namespace, routeSha1.Name, err)
+	}
+	defer func() {
```
Done.

gcs278 commented on test/e2e/route_upgradeable_test.go:
```diff
@@ -0,0 +1,95 @@
+	t.Log("resolving SHA1 route issue to remove admin gate")
+	certSha256, keySha256, err := generateClientCertificate(ca, caKey, "test", x509.SHA256WithRSA)
+	if err != nil {
+		t.Fatalf("failed to generate SHA256 client certificate: %v", err)
```
Done.

Miciah replied:
I think you missed this one.

gcs278 commented on test/e2e/util_test.go:
```diff
@@ -843,3 +845,32 @@ func createNamespace(t *testing.T, name string) *corev1.Namespace {
+// waitForAdminGate waits for an admin gate specified by adminGateKey
+// to either exist or not exist based on the expectExists argument.
+func waitForAdminGate(t *testing.T, adminGateKey string, expectExists bool, interval, timeout time.Duration) error {
+	err := wait.PollUntilContextTimeout(context.Background(), interval, timeout, false, func(ctx context.Context) (bool, error) {
```
Done.

gcs278 commented on test/e2e/util_test.go:
```diff
@@ -843,3 +845,32 @@ func createNamespace(t *testing.T, name string) *corev1.Namespace {
+		adminGateConfigMap := &corev1.ConfigMap{}
+		adminGateNamespacedName := types.NamespacedName{Name: routeupgradeable.AdminGatesConfigMapName, Namespace: operatorcontroller.GlobalMachineSpecifiedConfigNamespace}
+		if err := kclient.Get(ctx, adminGateNamespacedName, adminGateConfigMap); err != nil {
+			t.Logf("failed to get admin-gates ConfigMap %s/%s: %v", adminGateNamespacedName.Namespace, adminGateNamespacedName.Name, err)
```
Done.

gcs278 commented on test/e2e/util_test.go:
```diff
@@ -843,3 +845,32 @@ func createNamespace(t *testing.T, name string) *corev1.Namespace {
+
+		if adminGateConfigMap.Data == nil && !expectExists {
+			return true, nil
+		}
```
Ah yes, good point. Done.

gcs278 commented on test/e2e/util_test.go:
```diff
@@ -843,3 +845,32 @@ func createNamespace(t *testing.T, name string) *corev1.Namespace {
+		return false, nil
+	})
+	if err != nil {
+		return fmt.Errorf("failed to observe admin gate %q as %t: %v", adminGateKey, expectExists, err)
```
Done.

gcs278 reviewed (commented):
Thanks for the quick and detailed review. I addressed all of your comments.

frobware commented:
FAIL: TestAll/parallel/TestRouteMetricsControllerOnlyNamespaceSelector (77.93s)

/test e2e-aws-operator

frobware commented:
I took another look over the updates.

/lgtm

gcs278 commented:
@frobware @Miciah I removed the new util_tls_test.go file, and split the parts of the PR into more commits that changed existing E2E functions to hopefully make backports slightly less confusing:
https://github.com/openshift/cluster-ingress-operator/compare/abe6d326319212946a541928818c3fbc1a783176..8c22701c4fb719aa0fdeb6a54685fb7b8a741627

gcs278 commented:
`: [sig-network] pods should successfully create sandboxes by adding pod to network `:
/test e2e-aws-ovn-single-node

Miciah commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -725,6 +728,9 @@ func checkDefaultCertificate(secret *corev1.Secret, domain string) error {
 			return fmt.Errorf("certificate in secret %s/%s has legacy Common Name (CN) but has no Subject Alternative Name (SAN) for domain: %s", secret.Namespace, secret.Name, domain)
 		}
+		if cert.SignatureAlgorithm == x509.SHA1WithRSA || cert.SignatureAlgorithm == x509.ECDSAWithSHA1 {
+			return fmt.Errorf("certificate in secret %s/%s has weak SHA1 signature algorithm: %s\nsee https://docs.openshift.com/container-platform/4.16/release_notes/ocp-4-16-release-notes.html#ocp-4-16-networking for more details", secret.Namespace, secret.Name, domain)
```
[`SignatureAlgorithm` has a `String` method](https://cs.opensource.google/go/go/+/refs/tags/go1.21.7:src/crypto/x509/x509.go;l=244), so it'll print a string value; example: "One or more managed resources are not upgradeable: certificate in secret / has weak SHA1 signature algorithm: ECDSA-SHA1".

gcs278 replied:
Ah that's the missing link. I was confused playing around in Go Playground with these consts and it giving me errors that it's not the right type, but I didn't copy the String method. Added back in.

Miciah commented on pkg/operator/controller/route-metrics/controller.go:
```diff
@@ -144,6 +135,16 @@ func (r *reconciler) routeToIngressController(context context.Context, obj clien
+// Config holds all the configuration that must be provided when creating the
+// controller.
+type Config struct {
+	// Namespace specifies the namespace where Ingress Controllers reside.
```
```suggestion
	// Namespace specifies the namespace where IngressControllers reside.
```

gcs278 replied:
Done.

gcs278 commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -725,6 +728,10 @@ func checkDefaultCertificate(secret *corev1.Secret, domain string) error {
 		}
+		switch cert.SignatureAlgorithm {
+		case x509.SHA1WithRSA, x509.ECDSAWithSHA1:
+			return fmt.Errorf("certificate in secret %s/%s has weak SHA1 signature algorithm.\nsee https://docs.openshift.com/container-platform/4.16/release_notes/ocp-4-16-release-notes.html#ocp-4-16-networking for more details.", secret.Namespace, secret.Name)
```
Done.

Miciah replied:
You fixed it before I even submitted my review!  (Hm, maybe when I replied to another comment, GitHub pushed all my pending review comments?)

gcs278 replied:
Wow that wacky, not sure what happened.

wking commented on pkg/operator/controller/ingress/status.go:
```diff
@@ -725,6 +728,10 @@ func checkDefaultCertificate(secret *corev1.Secret, domain string) error {
 		}
+		switch cert.SignatureAlgorithm {
+		case x509.SHA1WithRSA, x509.ECDSAWithSHA1:
+			return fmt.Errorf("certificate in secret %s/%s has weak SHA1 signature algorithm: %s\nsee https://docs.openshift.com/container-platform/4.16/release_notes/ocp-4-16-release-notes.html#ocp-4-16-networking for more details", secret.Namespace, secret.Name, cert.SignatureAlgorithm)
```
Did you mean to link https://docs.openshift.com/container-platform/4.16/release_notes/ocp-4-16-release-notes.html#ocp-4-16-sha-haproxy-support-removed_release-notes ?  I don't see any `ocp-4-16-networking` anchors in the release notes at the moment, and while I do see things like `ocp-4-16-networking_release-notes`, I don't see a benefit to linking there vs. linking the specific HAProxy vs. SHA-1 section.

gcs278 replied:
Ah good catch. I could have sworn I tested this, I wonder if it changed recently. I'll use the specific anchor, and I guess we'll hope that doesn't change again. Done.

wking replied:
yeah, looks like openshift/openshift-docs@bdd9453e84a789d6ef2 brought in the more-specific anchor in April, and openshift/openshift-docs@0d46a75d67313d9fd5f01a80ff60d38487fec1ec broke your old anchor a few days ago 😅

Miciah commented on pkg/operator/controller/route-upgradeable/controller.go:
```diff
@@ -0,0 +1,285 @@
+	AdminGatesConfigMapName = "admin-gates"
+
+	// AdminAcksConfigMapName is the name of the admin-acks config map.
+	AdminAcksConfigMapName = "admin-acks"
```
Ah, I see https://github.com/openshift/cluster-ingress-operator/compare/e89e1f5d66e46c45133ac518489be35c3bff2da7..8c22701c4fb719aa0fdeb6a54685fb7b8a741627 removes `AdminAcksConfigMapName` as it wasn't used anywhere.

Miciah commented on pkg/operator/controller/route-upgradeable/controller.go:
```diff
@@ -0,0 +1,272 @@
+		GenericFunc: func(genericEvent event.GenericEvent) bool { return false },
+	}
+
+	toAdminGateConfigMap := func(ctx context.Context, _ client.Object) []reconcile.Request {
```
```suggestion
	toAdminGateConfigMap := func(_ context.Context, _ client.Object) []reconcile.Request {
```

gcs278 replied:
Done

Miciah commented:
Thanks for the quick updates!
/approve
/lgtm

Miciah commented:
This is fairly low risk.  The changes are scoped to a small refactoring in the route-metrics controller, a new route-upgradeable controller, and a new upgradeability check, all of which have good test coverage.  

/label backport-risk-assessed

wking commented:
[Flake][1]?

```
=== NAME  TestAll/parallel/TestIngressControllerRouteSelectorUpdateShouldClearRouteStatus
    panic.go:522: deleted ingresscontroller ic-route-selector-test
```

/test e2e-azure-operator

[1]: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_cluster-ingress-operator/1014/pull-ci-openshift-cluster-ingress-operator-release-4.15-e2e-azure-operator/1800626575222247424#1:build-log.txt%3A284-285

gcs278 commented:
@wking Yea it's caused by quite a simple problem and not related to this PR:
```
    router_status_test.go:134: failed to update ingresscontroller: Operation cannot be fulfilled on ingresscontrollers.operator.openshift.io "ic-route-selector-test": the object has been modified; please apply your changes to the latest version and try again
```
I'll make a bug eventually so NE can track this flake.

ShudiLi commented:
/label qe-approved
tested it with 4.15.0-0.ci.test-2024-06-12-014748-ci-ln-hmf17h2-latest(the upgrade test will be done after the PRs are merged)

```
1.
% oc get clusterversion
NAME      VERSION                                                   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.15.0-0.ci.test-2024-06-12-014748-ci-ln-hmf17h2-latest   True        False         18m     Cluster version is 4.15.0-0.ci.test-2024-06-12-014748-ci-ln-hmf17h2-latest

2.
% oc -n openshift-ingress-operator -c ingress-operator rsync ingress-operator-c89797b65-7kqqc:/usr/bin/ingress-operator .
receiving file list ... done
ingress-operator

sent 38 bytes  received 95493821 bytes  7639508.72 bytes/sec
total size is 95470416  speedup is 1.00

3.
% go version -m ingress-operator | grep github.com/openshift/api 
	dep	github.com/openshift/api	v3.9.1-0.20190924102528-32369d4db2ad+incompatible
	=>	github.com/openshift/api	v0.0.0-20240522145529-93d6bda14341
```

ShudiLi commented:
/test e2e-gcp-operator

ShudiLi commented:
/label cherry-pick-approved
